### ISSUE TITLE
CSP headers prevents the extension from wrapping JS APIs

### ISSUE THREAD

**Issue by polcak** [polcak/jsrestrictor#25]
CSP like `Content-Security-Policy: default-src 'self';` does not allow the wrapping to be executed. So far confirmed in FF.

I see several options that can be applied:

* `window.eval` in FF
* Script tags with data encoded URLs: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs

In both cases, we need to make sure that the wrapping code is executed.

**Comment by polcak** [comment]
It seems that this bug is Firefox-only.

**Comment by polcak** [comment]
`window.eval`: EvalError: call to eval() blocked by CSP

**Comment by polcak** [comment]
Data encoded URLs are also not an option: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src

**Comment by polcak** [comment]
This is known Firefox bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1267027, unfortunately it does not seem that they are making any progress lately.

**Comment by polcak** [comment]
A workaround could be to interact with page objects directly using https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts. However, that requires rewrite of the current code. Is window.wrappedJSObject supported in Chrome?

**Comment by matogolf** [comment]
Most probably not. I didn't find any solid evidence that window.wrappedJSObject is supported in Chrome.

**Comment by polcak** [comment]
It seems that Firefox needs wrapping such as:

```
  var text = "var f = () => 100; exportFunction(f, window.Date, {defineAs:'now'});";
  eval(text);
```

However, such wrapping is incompatible with our current code.

**Comment by polcak** [comment]
Proof of concept page https://www.fit.vutbr.cz/~ipolcak/jsr.php

**Comment by polcak** [comment]
JSR from c9a9b8b5a225cb5dc40f4156b0c0c84698bbeb5c is able to wrap the code but there are some problems with returning code that cannot be accessed by page script.

**Comment by polcak** [comment]
This will be fixed in imminent 0.5 using the code by G. Maone
